route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,238 @@ namespace route_following_plugin
+        if(maneuvers.size() < route_shortest_path.size()){
+            double target_speed_in_lanelet = findSpeedLimit(route_shortest_path.back());
+            double start_dist = wm_->routeTrackPos(route_shortest_path.back().centerline2d().front()).downtrack;
+            double end_dist = wm_->routeTrackPos(route_shortest_path.back().centerline2d().back()).downtrack; 
+            maneuvers.push_back(composeLaneFollowingManeuverMessage(start_dist, end_dist, prev_mvr_end_speed, target_speed_in_lanelet, route_shortest_path.back().id(),ros::Time::now()));
+        }
+        ROS_DEBUG_STREAM("Maneuver plan along route successfully generated");
+        return maneuvers;
+    }
 
+    bool RouteFollowingPlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp)
+    { 
+        if(latest_maneuver_plan_.empty()){
+            ROS_ERROR_STREAM("A maneuver plan has not been generated");
+            return false;
         }
-        resp.new_plan.maneuvers.push_back(
-            composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-            speed_progress,shortest_path[last_lanelet_index].id(),
-            shortest_path[last_lanelet_index].id(),time_progress,time_req_to_stop));
         
-        ROS_DEBUG_STREAM("Done Loop: approaching_route_end: " << approaching_route_end);
-        if(approaching_route_end){
-            ros::Time end_time = ros::Time::now();
-            if (resp.new_plan.maneuvers.size() > 0) {
-                end_time = resp.new_plan.maneuvers.back().lane_following_maneuver.end_time;
+        double current_downtrack = wm_->routeTrackPos(current_loc_).downtrack;
+        
+        //Return the set of maneuvers which intersect with min_plan_duration
+        int i=0;
+        double planned_time = 0.0;
+        while(planned_time < min_plan_duration_ && i < latest_maneuver_plan_.size()){
+            //Ignore plans for distance already covered
+            if(GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i], end_dist) < current_downtrack){
+                ++i;
+                continue;
             }
+            double maneuver_start_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i], start_time).toSec();
+            double maneuver_end_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i],end_time).toSec();
+            planned_time += maneuver_end_time - maneuver_start_time;
             
-            resp.new_plan.maneuvers.push_back(
-                composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-                speed_progress,shortest_path[last_lanelet_index].id(),
-                shortest_path[last_lanelet_index].id(),end_time,time_req_to_stop)); 
+            resp.new_plan.maneuvers.push_back(latest_maneuver_plan_[i]);
+            ++i;
         }
+        
         if(resp.new_plan.maneuvers.size() == 0)
         {
             ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
         }
```

Comment:
Should there be a return false here? @kjrush What is the desired behavior if planning is not successful? Return false or just return an empty list.